### PR TITLE
Add kernel time overlap validation test

### PR DESCRIPTION
**Pull request by nicksunka**
## Summary

This PR adds a profiler validation test to detect overlapping Spyre device-operation events in exported Chrome trace files.

The validation was initially scoped to compute kernel events. Based on review feedback, it now checks complete (`ph == "X"`) events from the Spyre device-operation categories currently relevant to this validation:

* `kernel`
* `gpu_memcpy`
* `gpu_memset`

CPU, runtime, annotation, driver, and profiler span events are not included in the overlap timeline.

The test:

* profiles a real Spyre workload using `ProfilerActivity.PrivateUse1`
* selects complete (`ph == "X"`) events from `kernel`, `gpu_memcpy`, and `gpu_memset`
* validates selected event timestamps and durations before checking overlaps
* requires timestamps to be finite numeric values
* requires durations to be finite and positive
* includes the event name, timestamp, duration, and end time in overlap diagnostics
* sorts selected device events by their start timestamp
* checks overlaps on the global Spyre device timeline for this validation
* flags kernel-to-kernel overlaps
* flags kernel-to-memory overlaps involving `gpu_memcpy` or `gpu_memset`
* does not flag memory-to-memory overlaps
* calculates the actual overlap interval using both event end times
* requires at least two selected Spyre device events so traces without relevant device activity cannot pass
* reports details for the first ten detected overlaps to help with debugging
* currently treats kernel-memory overlaps as invalid and flags them
* keeps the existing nested-kernel assumption: nested kernel intervals are treated as overlaps because AIU kernel events are currently expected to represent independent execution intervals rather than nested parent-child function execution

Stream-aware overlap validation is intentionally deferred to a follow-up through AIU Trace Analyzer. Trace Analyzer already provides per-stream handling, which will allow the validation to avoid false positives from overlaps across different streams. The follow-up work will be tracked under #3541.

Synthetic coverage includes:

* non-overlapping kernel events
* intentionally overlapping kernel events
* the expected overlapping kernel pair and overlap interval
* kernel-memory overlap using a `gpu_memcpy` event
* selection of a non-overlapping `gpu_memset` event
* memory-to-memory overlap being ignored
* zero-duration device events
* device events with a missing timestamp
* device events with a missing duration

The synthetic overlap duration assertion is expressed in trace time units rather than assuming a specific physical unit.

The hardware validation remains registered as a `mandatory_success` profiler validation in the Torch-Spyre test configuration.

Related issue - #3538

### Local

The latest review changes were validated with:

* focused synthetic overlap tests
* `git diff --check`
* Ruff format check
* Ruff check

Focused synthetic test result:

```text
2 passed in 1.24s
```

The final local formatting and lint checks also passed successfully.

### Trace inspection

The exported Spyre hardware trace was inspected before broadening the event filter.

Complete (`ph == "X"`) events included categories such as:

* `kernel`
* `gpu_memcpy`
* `gpu_memset`
* `privateuse1_driver`
* `privateuse1_runtime`
* `cpu_op`
* `user_annotation`
* `Trace`

The overlap validation therefore selects the device-operation categories `kernel`, `gpu_memcpy`, and `gpu_memset` rather than checking every complete event in the Chrome trace.

The inspected hardware trace contained 15 selected device-operation events and no overlaps under the updated validation scope.

### Earlier development validation

Before integrating the original kernel-overlap validation into the profiler pytest suite, the kernel-only overlap logic was manually tested against:

* 4 FMS Granite profiler traces — PASS
* clean synthetic trace — PASS
* synthetic trace with an intentional kernel overlap — FAIL
* Roberta trace with known kernel overlaps — FAIL

These checks were performed during the original kernel-only development phase.

### Spyre hardware pod

Validated on:

`nikhilsunka-spyre-dev-pf`

Using the prebuilt Torch-Spyre x86 image:

`icr.io/ai_sw_accel/2.0/torch-spyre:amd64`

Environment:

`PyTorch 2.13.0+cpu`

`torch_spyre` imported successfully, Spyre was available, and Spyre profiler activity was available through `ProfilerActivity.SPYRE`.

Validation test:

`tests/profiler/test_spyre_profiler.py::test_kernel_time_overlap`

Final result on the latest review changes:

```text
1 passed in 7.49s
```

The exported trace contained Spyre kernel, memcpy, and memset device events, and no kernel-to-kernel or kernel-to-memory overlaps were detected for this workload.

The profiler emitted a non-failing Kineto `[memory]` address-stability warning during collection, but the validation test completed successfully.

## Follow-up

The current PR intentionally keeps the overlap validation on the global Spyre device timeline.

The next step is to enable the overlap validation through AIU Trace Analyzer under #3541. This will reuse the analyzer's existing per-stream handling so overlaps across different streams are not incorrectly reported as failures.

Once stream IDs are propagated consistently for both kernel and memory events, the Trace Analyzer integration can apply the overlap validation within the appropriate stream.

--
Updated the profiler config to use the renamed overlap tests:
test_find_device_overlaps and test_find_device_overlaps_invalid_events.

I also validated the config with run_test.sh on the Spyre 2.13 environment:
9 passed, 4 skipped, overall exit code 0.

Attached the successful run output below.

Screenshot From 2026-08-25 09-00-11

### CODE REVIEW

File: tests/profiler/test_spyre_profiler.py [hunk 1]
```diff
@@ -366,3 +366,82 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    sorted_events = sorted(kernel_events, key=lambda event: event["ts"])
+    overlaps = []
+
+    for first_index, first_event in enumerate(sorted_events):
+        first_end_time = first_event["ts"] + first_event["dur"]
+
+        for second_event in sorted_events[first_index + 1 :]:
+            if second_event["ts"] >= first_end_time:
+                break
+            overlaps.append((first_event, second_event))
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: Use positive-duration interval semantics.

A zero-duration or negative-duration event passes the current filter. A zero-duration second event inside another interval is then reported as an overlap. Line 437 also overstates the overlap when the second event ends before the first event.

Filter for dur > 0. Calculate the intersection duration with both end times.

File: tests/profiler/test_spyre_profiler.py [hunk 2]
```diff
@@ -366,3 +366,82 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    if overlaps:
+        overlap_details = []
+
+        for first_event, second_event in overlaps[:10]:
+            first_end_time = first_event["ts"] + first_event["dur"]
+            overlap_time = first_end_time - second_event["ts"]
+            overlap_details.append(
+                f"{first_event.get('name', 'unknown')} overlaps "
+                f"{second_event.get('name', 'unknown')} by "
+                f"{overlap_time:.3f} us"
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: https://github.com/torch-spyre/torch-spyre/pull/3583/changes#r3730113986

File: tests/profiler/test_spyre_profiler.py [hunk 3]
```diff
@@ -366,3 +366,82 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    """Return compute kernel events and any overlapping event pairs."""
+    kernel_events = [
+        event
+        for event in events
+        if isinstance(event, dict)
+        and event.get("ph") == "X"
+        and event.get("cat") == "kernel"
+        and isinstance(event.get("ts"), (int, float))
+        and isinstance(event.get("dur"), (int, float))
+    ]
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: isinstance(value, (int, float)) accepts Boolean values, and json.load accepts non-finite values such as NaN. A zero or negative dur does not represent a valid kernel interval. Require finite, non-Boolean timestamps and finite durations greater than zero.

File: tests/profiler/test_spyre_profiler.py [hunk 4]
```diff
@@ -366,3 +366,82 @@ def test_memory_timeline_no_id_cpu(self) -> None:
 
         for event in expected:
             self.assertTrue(event in actual, f"event: {event} was not found in actual.")
+
+
+def _find_kernel_overlaps(events):
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:372]: Issue #3538 calls for synthetic coverage: clean trace (PASS) and synthetic overlap trace (FAIL), plus Granite/Roberta cases. Right now only a live Spyre workload is exercised, so _find_kernel_overlaps is never proven to fail when overlaps exist. Please add a small unit test that feeds synthetic traceEvents into _find_kernel_overlaps (one non-overlapping, one overlapping) and asserts PASS/FAIL. The hardware test can stay for end-to-end coverage.

File: tests/profiler/test_spyre_profiler.py [hunk 5]
```diff
@@ -366,3 +366,82 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+        and event.get("ph") == "X"
+        and event.get("cat") == "kernel"
+        and isinstance(event.get("ts"), (int, float))
+        and isinstance(event.get("dur"), (int, float))
+    ]
+
+    sorted_events = sorted(kernel_events, key=lambda event: event["ts"])
+    overlaps = []
+
+    for first_index, first_event in enumerate(sorted_events):
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:412]: Overlap is checked globally across every cat == "kernel" event, not per tid/stream/device. Is that intentional for Spyre (single serial compute stream in the trace)? If kernels on different tids can legally overlap, this will false-fail. Please confirm the intended scope, or group by tid (or equivalent) before comparing intervals.

File: tests/profiler/test_spyre_profiler.py [hunk 6]
```diff
@@ -366,3 +366,82 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+
+def _find_kernel_overlaps(events):
+    """Return compute kernel events and any overlapping event pairs."""
+    kernel_events = [
+        event
+        for event in events
+        if isinstance(event, dict)
+        and event.get("ph") == "X"
+        and event.get("cat") == "kernel"
+        and isinstance(event.get("ts"), (int, float))
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: Kernel events with missing/non-numeric ts/dur are silently filtered out. That can hide malformed events instead of failing. Prefer asserting that every `ph == "X" / cat == "kernel"` event has valid numeric ts and dur (and maybe dur >= 0) before the overlap check.

File: tests/profiler/test_spyre_profiler.py (hunk 2 above)
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py]: The overlap message hard-codes us. Please use the trace’s displayTimeUnit (if present) or phrase it as “trace time units” so the failure text isn’t misleading.

File: tests/profiler/test_spyre_profiler.py (hunk 5 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:412]: Good point. I've reached out to Priyanka to confirm whether the Spyre kernel events for this validation should be treated as a single global compute stream or scoped by tid/stream. For now, I've left the current global behavior unchanged until I have clarification on the expected semantics.

Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:412]: Confirmed with Priyanka. Overlap across different streams can be valid, but multi-stream support and the stream identifier exposed in the profiler trace still need confirmation. She’s okay with keeping the current global behavior for this PR, so I’ve left it unchanged for now pending confirmation from Tuan.

File: tests/profiler/test_spyre_profiler.py (hunk 3 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Updated this to check for finite, non-Boolean timestamps and durations before doing the overlap check.

File: tests/profiler/test_spyre_profiler.py (hunk 4 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:372]: Added a small synthetic test with both a clean non-overlapping case and an intentional overlap case. I had also tested the overlap logic earlier during development against the 4 Granite traces and the Roberta trace with known overlaps.

File: tests/profiler/test_spyre_profiler.py (hunk 6 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Updated this as well. Kernel events are validated before the overlap check now instead of silently dropping ones with invalid timestamp or duration values.

File: tests/profiler/test_spyre_profiler.py (hunk 2 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Updated this to use trace time units instead of hard-coding us.

File: tests/profiler/test_spyre_profiler.py (hunk 1 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Thanks, updated this to require valid positive kernel durations and calculate the actual overlap using both event end times.

File: tests/profiler/test_spyre_profiler.py [hunk 7]
```diff
@@ -366,3 +367,125 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+            isinstance(timestamp, (int, float))
+            and not isinstance(timestamp, bool)
+            and math.isfinite(timestamp)
+        ), "Kernel event must have a finite numeric timestamp"
+
+        assert (
+            isinstance(duration, (int, float))
+            and not isinstance(duration, bool)
+            and math.isfinite(duration)
+            and duration > 0
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:399]: Requiring duration > 0 here means test_kernel_time_overlap also fails on zero-duration kernels, which is a separate validation concern (see #3274 / zero ts-dur work). Consider either (a) documenting that this helper intentionally rejects zero-dur events as invalid input, or (b) separating “malformed event” checks from “overlap” checks so a zero-dur bug doesn’t look like an overlap failure. If keeping dur > 0, please include the event name in the assertion message for easier debugging.

File: tests/profiler/test_spyre_profiler.py [hunk 8]
```diff
@@ -366,3 +367,125 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+
+        for second_event in sorted_events[first_index + 1 :]:
+            if second_event["ts"] >= first_end_time:
+                break
+            overlaps.append((first_event, second_event))
+
+    return sorted_events, overlaps
+
+
+def test_find_kernel_overlaps():
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:423]: Nice addition for #3538. Please also cover the new validation path, e.g. dur == 0 / missing ts should raise AssertionError. Right now those asserts are untested.

File: tests/profiler/test_spyre_profiler.py [hunk 9]
```diff
@@ -13,18 +13,19 @@
...
 import json
-import pytest
+import math
 import unittest
+
+import pytest
 import torch
 import torch.nn.functional as F
-from torch.profiler import profile, ProfilerActivity, _memory_profiler
+from torch.profiler import ProfilerActivity, _memory_profiler, profile
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:22]: nit: Please avoid unrelated import reordering in this PR so the diff stays focused on the overlap test. Keep import cleanups for a separate change unless required by ruff on the new code only.

File: tests/profiler/test_spyre_profiler.py [hunk 10]
```diff
@@ -366,3 +367,125 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    for event in events:
+        if not isinstance(event, dict):
+            continue
+        if event.get("ph") != "X" or event.get("cat") != "kernel":
+            continue
+
+        timestamp = event.get("ts")
+        duration = event.get("dur")
+
+        assert (
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:386]: nit: Please include identifying fields in the assertion message (e.g. name, ts, dur) so a failing hardware trace is debuggable without dumping the whole JSON.

File: tests/profiler/test_spyre_profiler.py (hunk 5 above)
Comment (Vikram123-byte) [reply tests/profiler/test_spyre_profiler.py:412]: Thanks for confirming with Priyanka. Consider a short comment above the overlap loop noting that overlaps are checked on a global kernel timeline for now, and per-stream scoping is deferred until stream IDs in the trace are confirmed.

File: tests/profiler/test_spyre_profiler.py (hunk 7 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:399]: Thanks, I kept the positive-duration check and documented it in the helper. I also added the kernel name, ts and dur to the assertion message for easier debugging.

File: tests/profiler/test_spyre_profiler.py (hunk 8 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:423]: Added coverage for both a zero-duration kernel and a missing ts case. Both now check that the helper raises AssertionError.

File: tests/profiler/test_spyre_profiler.py (hunk 9 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:22]: I checked this one and tried restoring the original import order, but Ruff flags it as unsorted and requires the current ordering, So I've kept the Ruff-friendly version

File: tests/profiler/test_spyre_profiler.py (hunk 10 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:386]: Updated this. The assertion message now includes the kernel name, ts, and dur to make debugging easier.

File: tests/profiler/test_spyre_profiler.py (hunk 5 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:412]: Added a short comment above the overlap loop explaining that it currently checks overlaps on a global kernel timeline, with per-stream scoping deferred until the stream IDs are confirmed.

File: tests/profiler/test_spyre_profiler.py [hunk 11]
```diff
@@ -366,3 +367,151 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    ]
+
+    clean_kernels, clean_overlaps = _find_kernel_overlaps(clean_events)
+    overlap_kernels, overlaps = _find_kernel_overlaps(overlap_events)
+
+    assert len(clean_kernels) == 2
+    assert clean_overlaps == []
+
+    assert len(overlap_kernels) == 2
+    assert len(overlaps) == 1
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:442]: Please also assert the overlapping pair (kernel_1 vs kernel_2) and optionally the overlap interval (5 trace time units here). That makes the synthetic test a stronger guard against algorithm regressions.

File: tests/profiler/test_spyre_profiler.py [hunk 12]
```diff
@@ -366,3 +367,151 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+
+
+def test_find_kernel_overlaps_invalid_events():
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:454]: Good addition. Consider one more case for missing dur (or dur=None) so both assert paths are explicitly covered. Optional: negative/NaN duration if you want parity with the math.isfinite / duration > 0 checks.

File: tests/profiler/test_spyre_profiler.py [hunk 13]
```diff
@@ -366,3 +367,151 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    with trace_file.open("r", encoding="utf-8") as trace:
+        trace_data = json.load(trace)
+
+    assert isinstance(trace_data, dict), "Trace JSON must be a dictionary"
+    assert "traceEvents" in trace_data, "Chrome trace is missing the 'traceEvents' key"
+
+    trace_events = trace_data["traceEvents"]
+    assert isinstance(trace_events, list), "'traceEvents' must contain a list"
+
+    kernel_events, overlaps = _find_kernel_overlaps(trace_events)
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:507]: With duration > 0 enforced in _find_kernel_overlaps, this hardware test will fail on malformed kernel intervals (zero/missing dur) before overlap is evaluated. That’s fine if intentional — please add a one-line note in the test docstring so CI failures aren’t mistaken for overlap bugs. (Zero-duration validation is tracked separately in #3274 / #3545.)

**Comment by Vikram123-byte** [comment]
@nicksunka one more issue I found: [#3538](https://github.com/torch-spyre/torch-spyre/issues/3538) listed Granite PASS and Roberta FAIL traces. PR says those were validated manually on pod; not in CI. If we want regression coverage for known-overlap traces in CI later, consider small checked-in fixture JSONs + a parametrized test.

File: tests/profiler/test_spyre_profiler.py (hunk 11 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:442]: Updated this to assert the overlapping kernel_1 / kernel_2 pair and verify the expected 5 trace-time-unit overlap as well.

File: tests/profiler/test_spyre_profiler.py (hunk 12 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:454]: Added a missing dur case as well so both validation paths are explicitly covered.

File: tests/profiler/test_spyre_profiler.py (hunk 13 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:507]: Updated the test docstring to make it clear that kernel interval validation happens before the overlap check.

File: tests/profiler/test_spyre_profiler.py [hunk 14]
```diff
@@ -366,3 +367,167 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    assert len(overlaps) == 1
+
+    first_event, second_event = overlaps[0]
+    assert first_event["name"] == "kernel_1"
+    assert second_event["name"] == "kernel_2"
+
+    first_end_time = first_event["ts"] + first_event["dur"]
+    second_end_time = second_event["ts"] + second_event["dur"]
+    overlap_time = min(first_end_time, second_end_time) - second_event["ts"]
+    assert overlap_time == 5
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:451]: Please use the same overlap-interval formula here as in `test_kernel_time_overlap (overlap_end - max(start1, start2))`. The current shortcut assumes pair order from the sorted sweep and will silently assert the wrong value if that invariant changes. Keeping one formula avoids drift between the unit test and the hardware failure path.

File: tests/profiler/test_spyre_profiler.py [hunk 15]
```diff
@@ -366,3 +367,167 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+            second_end_time = second_event["ts"] + second_event["dur"]
+
+            overlap_start = max(first_event["ts"], second_event["ts"])
+            overlap_end = min(first_end_time, second_end_time)
+            overlap_time = overlap_end - overlap_start
+
+            overlap_details.append(
+                f"{first_event.get('name', 'unknown')} overlaps "
+                f"{second_event.get('name', 'unknown')} by "
+                f"{overlap_time:.3f} trace time units"
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:527]: Please include ts/dur (and maybe ts+dur) for both events in the overlap failure message. On a real Spyre trace, name-only output is hard to map back to the Chrome JSON without dumping the whole file.

File: tests/profiler/test_spyre_profiler.py [hunk 16]
```diff
@@ -366,3 +367,167 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    with pytest.raises(AssertionError):
+        _find_kernel_overlaps(missing_timestamp_event)
+
+    with pytest.raises(AssertionError):
+        _find_kernel_overlaps(missing_duration_event)
+
+
+@pytest.mark.requires_spyre_profiler
+def test_kernel_time_overlap(tmp_path):
+    """Verify kernel intervals are valid and compute kernel events do not overlap."""
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:480]: Nit on wording: “intervals are valid” here only means finite ts + positive dur for overlap arithmetic (and ts == 0 is still allowed). Please say that explicitly so this isn’t confused with the zero-ts/dur profiler checks in #3545. Example: “Reject non-finite/non-positive kernel intervals, then assert no overlaps.”

File: tests/profiler/test_spyre_profiler.py (hunk 14 above)
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:451]: Can we add a couple of synthetic edge cases: (1) identical start timestamps, (2) nested interval, (3) three mutually overlapping kernels expecting 3 pairs? The current cases don’t exercise equal-ts or multi-pair reporting, which is where sweep-line bugs usually hide.

File: tests/profiler/test_spyre_profiler.py [hunk 17]
```diff
@@ -366,3 +367,167 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+
+
+def _find_kernel_overlaps(events):
+    """Return valid positive-duration kernel events and any overlapping event pairs."""
+    kernel_events = []
+
+    for event in events:
+        if not isinstance(event, dict):
+            continue
+        if event.get("ph") != "X" or event.get("cat") != "kernel":
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:379]: Confirming: Spyre chrome traces always emit compute as ph == "X" + cat == "kernel", and never put non-compute activity in that category when both ProfilerActivity.CPU and PrivateUse1 are enabled? If runtime / other activities can also land in kernel, we should tighten the filter (or document why this is safe). Global timeline is fine per Priyanka for now.

File: tests/profiler/test_spyre_profiler.py [hunk 18]
```diff
@@ -366,3 +367,167 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    with trace_file.open("r", encoding="utf-8") as trace:
+        trace_data = json.load(trace)
+
+    assert isinstance(trace_data, dict), "Trace JSON must be a dictionary"
+    assert "traceEvents" in trace_data, "Chrome trace is missing the 'traceEvents' key"
+
+    trace_events = trace_data["traceEvents"]
+    assert isinstance(trace_events, list), "'traceEvents' must contain a list"
+
+    kernel_events, overlaps = _find_kernel_overlaps(trace_events)
```
Comment (Vikram123-byte) [inline tests/profiler/test_spyre_profiler.py:507]: https://github.com/torch-spyre/torch-spyre/pull/3583/changes#r3749990133

**Review by Vikram123-byte — COMMENTED** [review]
Re-reviewed the latest push. The sweep-line overlap check looks correct for a global kernel timeline (touching endpoints are non-overlapping; nested/same-start cases are detected). Remaining comments are about consistency of the overlap-interval formula between the synthetic and hardware tests, failure-message debuggability, avoiding unrelated import churn in test_chrome_trace_is_valid_json, and tightening docstring/synthetic edge coverage

File: tests/profiler/test_spyre_profiler.py [hunk 19]
```diff
@@ -366,3 +367,167 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+
+    # Check overlaps on the global kernel timeline for now.
+    # Per-stream scoping is deferred until stream IDs in the trace are confirmed.
+    for first_index, first_event in enumerate(sorted_events):
+        first_end_time = first_event["ts"] + first_event["dur"]
+
+        for second_event in sorted_events[first_index + 1 :]:
+            if second_event["ts"] >= first_end_time:
+                break
+            overlaps.append((first_event, second_event))
```
Comment (lasch) [inline tests/profiler/test_spyre_profiler.py:418]: One minor comment that could be worth noting/documenting somewhere:

This overlap detection will flag technically legitimate parent-child function execution as an overlap.

Fully embedded overlap (regular fn-call) within single processing unit:
```
<----->
  <-->
```

Impossible within single processing unit (unless async/shared execution):
```
<----->
   <------>
```

As far as I'm aware, this fn-call pattern currently cannot happen with AIU kernel events, so the implementation would **correctly** flag any such overlapping spans as an issue.

Just need to keep an eye out for this in case it becomes possible in the future and then all of a sudden the fn-call case would no longer signal broken timings.

Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:581]: Good point, thanks Lars. I’ve added a note about this assumption next to the overlap logic and also mentioned it in the issue. For now, AIU kernel events aren’t expected to have nested parent-child execution, so fully nested kernel spans are treated as overlaps. If that changes in the future, we can revisit the validation.

**Comment by nicksunka** [comment]
Just pushed a small docs update based on Lars’s comment. I added a note about the nested parent-child case next to the overlap logic and also added it to #3538 so we have it there for future reference. No changes to the actual logic or test behavior.

File: tests/profiler/test_spyre_profiler.py (hunk 14 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Updated this to use the same overlap calculation as the hardware test, so they both handle it the same way.

File: tests/profiler/test_spyre_profiler.py (hunk 15 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py:748]: Updated the failure message to include ts, dur, and the end time for both events so it’s easier to map them back to the trace.

File: tests/profiler/test_spyre_profiler.py (hunk 16 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Updated the docstring to explicitly say it rejects non-finite timestamps and non-positive durations before checking for overlaps.

File: tests/profiler/test_spyre_profiler.py (hunk 14 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Thanks, I did consider these. I’d prefer to keep the synthetic tests focused on the overlap cases we’re covering in #3538. Identical start timestamps are being handled separately in #3539, and I’ve added a note about the nested case based on Lars’s comment. I think the current coverage should be enough for this PR.

File: tests/profiler/test_spyre_profiler.py (hunk 17 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Yeah, the traces I’ve tested use ph == "X" and cat == "kernel" for the compute events. I’ll double-check if any runtime or other non-compute events can also show up under that category before changing the filter. I’ll leave the global timeline as it is for now based on Priyanka’s feedback.

File: tests/profiler/test_spyre_profiler.py (hunk 18 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Yep, same as the filter comment above. I’ll confirm that first before changing anything here.

**Comment by nicksunka** [comment]
FYI, I’ve pushed the latest review changes and tested everything locally and on the Spyre pod. The CI checks are still red because of the CI issue that’s been going on since this morning. Once that’s sorted, I’ll rerun the checks and take a look at anything that comes up from my changes.

File: tests/profiler/test_spyre_profiler.py (hunk 17 above)
Comment (ppnaik1890) [reply tests/profiler/test_spyre_profiler.py]: I think we may have to add check overlap between a kernel and the subsequent mem event as well. Also, in future we are looking to add more events so do we need the category check or we can do the overlap check for all events on spyre?

File: tests/profiler/test_spyre_profiler.py [hunk 20]
```diff
@@ -366,3 +367,180 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+    first_start_time = first_event["ts"]
+    second_start_time = second_event["ts"]
+    first_end_time = first_start_time + first_event["dur"]
+    second_end_time = second_start_time + second_event["dur"]
+
+    overlap_start = max(first_start_time, second_start_time)
+    overlap_end = min(first_end_time, second_end_time)
+    overlap_time = overlap_end - overlap_start
+
+    assert overlap_time == 5
```
Comment (ppnaik1890) [inline tests/profiler/test_spyre_profiler.py:670]: this is is what duration value us or ms? will be good to add this as a comment.

File: tests/profiler/test_spyre_profiler.py [hunk 21]
```diff
@@ -366,3 +367,180 @@ def test_memory_timeline_no_id_cpu(self) -> None:
...
+
+
+def test_find_kernel_overlaps_invalid_events():
```
Comment (ppnaik1890) [inline tests/profiler/test_spyre_profiler.py]: Can we skip this test if we identify overlap across all events?

**Review by ppnaik1890 — CHANGES_REQUESTED** [review]

File: tests/profiler/test_spyre_profiler.py (hunk 17 above)
Comment (lasch) [reply tests/profiler/test_spyre_profiler.py]: @ppnaik1890 mem events can overlap with kernel events in arbitrary ways, I think. At least in the past, we've seen DmaO events start pretty much simultaneously with the kernel and then ended later. A common pattern of posting the data transfer op early so that it's set up and ready once the kernel has produced the data. (Unless, there's some other high-level constraint that would require non-overlap between mem and compute ops)

Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Updated this based on the check now covers complete kernel, gpu_memcpy, and gpu_memset events, and catches kernel-kernel and kernel-memory overlaps. Memory-memory overlaps are left alone.

I’ve kept the global timeline for this PR. We’ll handle the stream-aware part separately through the Trace Analyzer/Acelyzer follow-up, so we don’t end up flagging valid overlaps happening on different streams.

File: tests/profiler/test_spyre_profiler.py (hunk 21 above)
Comment (nicksunka) [reply tests/profiler/test_spyre_profiler.py]: Kept this test as discussed. It now validates invalid timestamp/duration data for the selected Spyre device events.

**Review by ppnaik1890 — DISMISSED** [review]
lgtm

**Comment by ppnaik1890** [comment]
@ashokponkumar @anubhavjana can you please give next round of review. Thanks!
cc @seepgoel

**Comment by seepgoel** [comment]
@JRosenkranz @ani300 can you please review this PR ?

File: tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml [hunk 22]
```diff
@@ -20,3 +20,8 @@ test_suite_config:
             - TestMemoryProfilerTimeline::test_memory_timeline_no_id_spyre
             - TestMemoryProfilerTimeline::test_memory_timeline_no_id_cpu
           mode: mandatory_success
+        - names:
+            - test_find_kernel_overlaps
```
Comment (JRosenkranz) [inline tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml]: should these be `test_find_device_overlaps` and `test_find_device_overlaps_invalid_events`?

Comment (nicksunka) [reply tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml]: Yes,  Those should be test_find_device_overlaps and test_find_device_overlaps_invalid_events. I’ve already updated the profiler config locally and I’ll push the fix after I rerun the config validation with run_test.sh once the Spyre pod is available again. Thanks!

File: tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml [hunk 23]
```diff
@@ -20,3 +20,8 @@ test_suite_config:
             - TestMemoryProfilerTimeline::test_memory_timeline_no_id_spyre
             - TestMemoryProfilerTimeline::test_memory_timeline_no_id_cpu
           mode: mandatory_success
+        - names:
```
Comment (JRosenkranz) [inline tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml:23]: These bare module-level names don't match the framework, which only recognizes Class::method entries — so this block matches nothing and attaches no mode. The tests already run as mandatory_success via unlisted_test_mode, so either drop the block or move them into a TestCase and register as Class::method

Comment (anubhavjana) [reply tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml:23]: @JRosenkranz I had added a feature to allow for bare methods also for the test framework, if I vaguely remember 😄 

I have asked @nicksunka to run the profiler config locally using the framework once to validate ahd share the output in the PR.

File: tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml (hunk 22 above)
Comment (nicksunka) [reply tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml]: I’ve updated these to test_find_device_overlaps and test_find_device_overlaps_invalid_events and pushed the change.

File: tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml (hunk 23 above)
Comment (nicksunka) [reply tests/configs/torch_spyre_tests/test_spyre_profiler_config.yaml:23]: Confirmed this works with the bare test names. I reran the profiler config through run_test.sh on the Spyre 2.13 environment after the update — 9 passed, 4 skipped, exit code 0. I’ve added the validation output to the PR as well.

**Review by anubhavjana — APPROVED** [review]
